hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/MultipleSparkJobExecutionStrategy.java

```diff
@@ -205,12 +207,26 @@ public MultipleSparkJobExecutionStrategy(HoodieTable table, HoodieEngineContext
               .withSpillableMapBasePath(config.getSpillableMapBasePath())
               .build();
 
-          HoodieTableConfig tableConfig = table.getMetaClient().getTableConfig();
-          recordIterators.add(getFileSliceReader(baseFileReader, scanner, readerSchema,
-              tableConfig.getPayloadClass(),
-              tableConfig.getPreCombineField(),
-              tableConfig.populateMetaFields() ? Option.empty() : Option.of(Pair.of(tableConfig.getRecordKeyFieldProp(),
-                  tableConfig.getPartitionFieldProp()))));
+          if (!StringUtils.isNullOrEmpty(clusteringOp.getDataFilePath())) {
+            HoodieFileReader<? extends IndexedRecord> baseFileReader = HoodieFileReaderFactory.getFileReader(table.getHadoopConf(), new Path(clusteringOp.getDataFilePath()));
+            HoodieTableConfig tableConfig = table.getMetaClient().getTableConfig();
+            recordIterators.add(getFileSliceReader(baseFileReader, scanner, readerSchema,
+                tableConfig.getPayloadClass(),
+                tableConfig.getPreCombineField(),
+                tableConfig.populateMetaFields() ? Option.empty() : Option.of(Pair.of(tableConfig.getRecordKeyFieldProp(),
+                    tableConfig.getPartitionFieldProp()))));
+          } else {
+            // Since there is no base file, fall back to reading log files
+            Iterable<HoodieRecord<? extends HoodieRecordPayload>> iterable = () -> scanner.iterator();
```

Comment:
Functionality looks good. But what do you think reorganizing this a little? Here is what I'm thinking:

Change HoodieFileSliceReader#getFileSliceReader method to take Option[HoodieBaseFileReader].  This whole logic can be embedded inside that method (Introduce new methods if needed).  

Makes it easy to reuse code if there are other places that need to read FileSlices. Please try and let me know if you think this is reasonable.

Reply:
+1 was about to suggest the same. We are nearing the release though. So, I would suggest to time bound. If not, atleast file a tracking ticket.

Reply:
+1 refactored as suggested.